src/Orleans.Serialization/Utilities/VarIntReaderExtensions.cs

```diff
@@ -16,7 +17,7 @@ public static class VarIntReaderExtensions
         /// <param name="reader">The reader.</param>
         /// <returns>A variable-width integer.</returns>
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public static sbyte ReadVarInt8<TInput>(this ref Reader<TInput> reader) => ZigZagDecode((byte)reader.ReadVarUInt32());
+        public static sbyte ReadVarInt8<TInput>(this ref Reader<TInput> reader) => ZigZagDecode(checked((byte)reader.ReadVarUInt32()));
```

Comment:
Do the `checked` have a measurable perf-impact?
I'm asking because it [produces more code](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAMIR8AB14AbGFADKMgG68wMXAG4aNYgGYmpBkIYBvGgzNNdvHgwCy5ABSSIXAOYMF2SRxgBKBgF4APgZ7KwwfDy8YDWpzCwYw21JHZzdI7z8ghjAACxgwAGsYABN7UJ4IzwyYgF8gA=) which may change inlining-behavior, code-flow (potential branches), etc.

TBH I don't have any better idea, except a manual check if within target-range but that's not cheaper.

Reply:
For the test I ran, the impact of the branch in ReadVarUInt32 at least came down to noise, but I'm unsure about the checked conversion cost. We could conceivably allow it to remain as a silent overflow, but I'd rather err on the side of caution.

I considered adding an attribute for fields which have been narrowed to instruct the code generator to opt in to overflow checks, but I'm not convinced of the value yet.

Reply:
RyuJIT handles checked conversions in a way that should generally not affect any inlining or branching (except for the actual jmp instruction which should be predicted never taken by the CPU).